### PR TITLE
localization

### LINKED ISSUE #53: [Feature Request]: Localization
### What should be implemented?

Allow localization of the GUI

### PR DESCRIPTION
**Pull request by JayPi4c**
## What type of PR is this? (keep all applicable)

- Feature

## Description

This PR implements a i18n for the gui. It has the option to select supported languages in the login window. The language will change at runtime in the login window and the translation will be applied to all other windows as well. In the login window, `StringBindings` have been used to allow real time translation after a language has been selected. Other windows do not offer the option to change the language and take the text from the localized messages file. This way all texts can be defined as keys in the `fxml` file and when a language is selected, all scenes will be displayed with the correct language. 

One known issue so far is, that the title for the login scene does not change according to the selected language. For this the login controller needs to obtain the `stage` object. For this I only have found some hacky ways. For example taking the root object and its parent is supposed to be the stage.  

Also the `comboBox` in the login window has only been added as an example. I would much more like it to be a little icon that upon clicking allows to select the language. But I could not figure out, how to load an icon and how to place it somewhere else in the login window, thus the current position and (non-existing) style.

## Related Tickets & Documents

- Closes #53

### CODE REVIEW

File: client/src/main/java/org/schlunzis/kurtama/client/util/I18nUtils.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    /**
+     * The current locale. Will be initially set to systems default ({@link Locale#getDefault()}).
+     */
+    @Getter
+    private static Locale locale;
+
+    static {
+        // setLocale(SUPPORTED_LOCALES.getFirst());
+        setLocale(Locale.getDefault());
```
Comment (Til7701) [inline client/src/main/java/org/schlunzis/kurtama/client/util/I18nUtils.java]: You could use the user settings to save the selection for the next time.

File: client/src/main/resources/lang/messages_de_DE.properties [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+main.lobby.button.join=Lobby beitreten
+main.lobby.label.name=Name:
+main.lobby.label.password=Passwort:
+main.lobby.prompt.searchField=Lobbies durchsuchen
+main.menu.button.logout=Abmelden
+main.menu.button.quit=Beenden
+main.menu.button.settings=Einstellungen
+title.lobby=Lobby
+title.login=Anmeldung
+title.main=Hauptmen�
```
Comment (Til7701) [inline client/src/main/resources/lang/messages_de_DE.properties]: There seems to be an encoding problem here.

**Review by Til7701 — CHANGES_REQUESTED** [review]

File: client/src/main/java/org/schlunzis/kurtama/client/util/I18nUtils.java (hunk 1 above)
Comment (JayPi4c) [reply client/src/main/java/org/schlunzis/kurtama/client/util/I18nUtils.java]: Done!

File: client/src/main/resources/lang/messages_de_DE.properties (hunk 2 above)
Comment (JayPi4c) [reply client/src/main/resources/lang/messages_de_DE.properties]: Done!

**Comment by JayPi4c** [comment]
I made all changes to make the i18n class to be more aligned with the spring style. But somehow the `setLocale` method seems broken. Whenever I remove the logging of the previous and the new Bundle, the change of bundle is not properly recognized and the `StringBindings` are not informed to change. Whatever I do, I can't find a reason for this. My guess - that I can't prove - is, that the `ObjectProperty` does not evaluate the old and new values as different and therefore, the listeners are not informed. But I could not find evidence for this...
Please help!

**Comment by Til7701** [comment]
The attribute `valid` in `ObjectPropertyBase` has to be set to true, before any listeners are informed of changes. That is done by calling the `get` method. Listeners apparently do not need that method to get the new value. So the following happens:

1. The value of `valid` is initially `true`
2. Call to `set`
3. Call to `markInvalid`
4. `valid` is set to `false`
5. Listeners are informed
6. UI updated
7. Second call to `set`
8. Value is replaced
9. `valid` is still `false`, so no events are fired in `markInvalid`

I don't know, what to do in this case. You could just insert any call to the `get` method. `toString` uses the `get` method. That's why the log is fixing the issue.

**Comment by JayPi4c** [comment]
It's really interesting that a get is required in order to validate the changes. Building on your findings I implemented a workaround for now. I also left a note and a todo to check this out later. Additionally the supported locales are not loaded at startup, so adding a new language is as easy as adding a new `messages_*.properties` file and fill it in with `key`-`value` pairs. And because we have the fallback `messages.properties`, the locale specific messages files don't need to be complete. Example:
```properties
main.lobby.label.name=Nom:
```
Can be added as `messages_fr_FR.properties` to the resource folder. The France locale will be picked up and the label will be displayed in french. All other texts will be shown in the fallback language, which is English for now.

**Review by Til7701 — APPROVED** [review]

**Pull request merged**